google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -341,10 +341,17 @@ private void processBatch(List<OutstandingMessage> batch) {
         // This should be a blocking flow controller and never throw an exception.
         throw new IllegalStateException("Flow control unexpected exception", unexpectedException);
       }
-      processOutstandingMessage(message.receivedMessage.getMessage(), message.ackHandler);
+      processOutstandingMessage(addDeliveryInfoCount(message.receivedMessage), message.ackHandler);
     }
   }
 
+  private PubsubMessage addDeliveryInfoCount(ReceivedMessage receivedMessage) {
+    return PubsubMessage.newBuilder(receivedMessage.getMessage())
+        .putAttributes(
+            "googclient_deliveryattempt", Integer.toString(receivedMessage.getDeliveryAttempt()))
```

Comment:
What happens if this key already exists?

Reply:
The value will be overwritten if the key already exists, so we should be ok here

Reply:
SGTM